certbot-apache/certbot_apache/configurator.py

```diff
@@ -280,9 +299,91 @@ def deploy_cert(self, domain, cert_path, key_path,
             a lack of directives
 
         """
-        # Choose vhost before (possible) enabling of mod_ssl, to keep the
-        # vhost choice namespace similar with the pre-validation one.
-        vhost = self.choose_vhost(domain)
+        if self.wildcard_domain(domain):
+            # Ask user which VHosts to support.
+            # Returned objects are guaranteed to be ssl vhosts
+            deploy_vhosts = self.choose_vhosts_wildcard(domain, create_ssl=True)
+            for vhost in deploy_vhosts:
+                self._deploy_cert(vhost, cert_path, key_path,
+                                  chain_path, fullchain_path)
+                if domain not in self.wildcard_vhosts.keys():
+                    self.wildcard_vhosts[domain] = list()
+                self.wildcard_vhosts[domain].append(vhost)
+        else:
+            vhost = self.choose_vhost(domain)
+            self._deploy_cert(vhost, cert_path, key_path, chain_path, fullchain_path)
+
+    def vhosts_for_wildcard(self, domain):
+        """
+        Get VHost objects for every VirtualHost that the user wants to handle
+        with the wildcard certificate.
+        """
+
+        # Collect all vhosts that match the name
+        matched = set()
+        for vhost in self.vhosts:
+            for name in vhost.get_names():
+                if self._in_wildcard_scope(name, domain):
+                    matched.add(vhost)
+
+        return list(matched)
+
+    def _in_wildcard_scope(self, name, domain):
+        """
+        Helper method for vhosts_for_wildcard() that makes sure that the domain
+        is in the scope of wildcard domain.
+
+        eg. in scope: domain = *.wild.card, name = 1.wild.card
+        not in scope: domain = *.wild.card, name = 1.2.wild.card
+        """
+        if len(name.split(".")) == len(domain.split(".")):
+            return fnmatch.fnmatch(name, domain)
+
+
+    def choose_vhosts_wildcard(self, domain, create_ssl=True):
+        """Prompts user to choose vhosts to install a wildcard certificate for"""
+
+        # Get all vhosts that are covered by the wildcard domain
+        vhosts = self.vhosts_for_wildcard(domain)
+
+        # Go through the vhosts, making sure that we cover all the names
+        # present, but preferring the SSL vhosts
+        filtered_vhosts = dict()
+        for vhost in vhosts:
+            for name in vhost.get_names():
+                if vhost.ssl:
+                    # Always prefer SSL vhosts
+                    filtered_vhosts[name] = vhost
+                elif name not in filtered_vhosts.keys():
+                    # Add if not in list previously
+                    filtered_vhosts[name] = vhost
```

Comment:
what happens if there aren't any? do we want to create a new one?

Reply:
Yeah, we do create one when appending the vhosts to the final list in the method.